### PR TITLE
Add variadic templates materials

### PR DESCRIPTION
**Pull request by kboyarinov**
Add variadics materials + numeration fix + SFINAE fixes

### CODE REVIEW

File: 12_Variadic_templates/12_00_problem/12_00_00_problem_statement.cpp [hunk 1]
```diff
@@ -0,0 +1,24 @@
+// How to implement execute function
```
Comment (rarutyun) [inline 12_Variadic_templates/12_00_problem/12_00_00_problem_statement.cpp]: Variadic_templates -> variadic_templates

File: 12_Variadic_templates/12_01_syntax/12_01_01_terms.cpp [hunk 2]
```diff
@@ -0,0 +1,5 @@
+
+template <typename... Args> // template parameter pack declaration
+void f1( Args... args ) { // function parameter pack declaration
+    f0(args...); // parameter pack expansion
+}
```
Comment (rarutyun) [inline 12_Variadic_templates/12_01_syntax/12_01_01_terms.cpp]: New line here and in other many places

File: 12_Variadic_templates/12_04_recursive_cut_off/12_04_00_recursive_cut_off.cpp [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+
+
+
+
+
+
+
+
+void execute() {}
```
Comment (rarutyun) [inline 12_Variadic_templates/12_04_recursive_cut_off/12_04_00_recursive_cut_off.cpp]: I believe we implement execute with one argument. I would propose the following:
remove line 43. Leave execute with Variadics as a problem statement without the solution and move it closely to sum.

Add comment to the very top. Something like: "// Solution for execute"

File: 13_template_meta_programming/13_00_compile_time_computations/13_00_00_constant_expressions.cpp [hunk 4]
```diff
@@ -1,17 +1,17 @@
-// What are constant expressions and where they may be used?
```
Comment (rarutyun) [inline 13_template_meta_programming/13_00_compile_time_computations/13_00_00_constant_expressions.cpp]: I believe you also changed the line ending for this file and below. Diff looks huge. I think we should revert that.

File: 12_Variadic_templates/12_01_syntax/12_01_00_introduction_into_syntax.cpp [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+struct Object {};
+
+int main() {
+
+    f0(1, 2., Object{});
+
+    // template <typename T1, typename T2, typename T3>
+    // void f0( T1 arg1, T2 arg2, T3 arg3 );
+
+    // void f0<int, float, A>(1, 2f, A{});
```
Comment (ivankochin) [inline 12_Variadic_templates/12_01_syntax/12_01_00_introduction_into_syntax.cpp]:
```suggestion
    // void f0<int, float, Object>(1, 2f, Object{});
```

File: 12_Variadic_templates/12_00_problem/12_00_00_problem_statement.cpp (hunk 1 above)
Comment (kboyarinov) [reply 12_Variadic_templates/12_00_problem/12_00_00_problem_statement.cpp]: Done

File: 12_Variadic_templates/12_04_recursive_cut_off/12_04_00_recursive_cut_off.cpp (hunk 3 above)
Comment (kboyarinov) [reply 12_Variadic_templates/12_04_recursive_cut_off/12_04_00_recursive_cut_off.cpp]: Done

File: 13_template_meta_programming/13_00_compile_time_computations/13_00_00_constant_expressions.cpp (hunk 4 above)
Comment (kboyarinov) [reply 13_template_meta_programming/13_00_compile_time_computations/13_00_00_constant_expressions.cpp]: Done

File: 12_Variadic_templates/12_01_syntax/12_01_01_terms.cpp (hunk 2 above)
Comment (kboyarinov) [reply 12_Variadic_templates/12_01_syntax/12_01_01_terms.cpp]: Done

File: 12_variadic_templates/12_00_problem/12_00_01_simple_solution.cpp [hunk 6]
```diff
@@ -0,0 +1,45 @@
+// Show execute implementation with several overloads
```
Comment (ivankochin) [inline 12_variadic_templates/12_00_problem/12_00_01_simple_solution.cpp]:
```suggestion
// Show `execute` implementation with several overloads
```

File: 12_variadic_templates/12_00_problem/12_00_02_variadic_solution.cpp [hunk 7]
```diff
@@ -0,0 +1,25 @@
+// Introduce variadic templates as a solution
```
Comment (ivankochin) [inline 12_variadic_templates/12_00_problem/12_00_02_variadic_solution.cpp]:
```suggestion
// Variadic templates as a solution
```

File: 12_variadic_templates/12_02_pack_expansion/12_02_00_simple_operation.cpp [hunk 8]
```diff
@@ -0,0 +1,13 @@
+#include <iostream>
+
+template <typename... Args>
+void g( Args... args ) {}
+
+template <typename... Args>
+void f( Args... args ) { // args ~= arg1, arg2, arg3, ...
```
Comment (ivankochin) [inline 12_variadic_templates/12_02_pack_expansion/12_02_00_simple_operation.cpp]:
```suggestion
void f( Args... args ) { // arg1, arg2, arg3, ...
```

File: 12_variadic_templates/12_02_pack_expansion/12_02_01_template.cpp [hunk 9]
```diff
@@ -0,0 +1,24 @@
+// print types
```
Comment (ivankochin) [inline 12_variadic_templates/12_02_pack_expansion/12_02_01_template.cpp]:
```suggestion
// Print types
```

File: 12_variadic_templates/12_02_pack_expansion/12_02_01_template.cpp [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+template <typename... Args>
+struct ObjectTypes {
+
+    using type1 = Object<Args...>;
+    using type2 = Object<int, Args...>;
+    using type3 = Object<Args..., int>;
+    using type4 = Object<int, Args..., int>;
+
+};
```
Comment (ivankochin) [inline 12_variadic_templates/12_02_pack_expansion/12_02_01_template.cpp]:
```suggestion
struct ObjectTypes {
    using type1 = Object<Args...>;
    using type2 = Object<int, Args...>;
    using type3 = Object<Args..., int>;
    using type4 = Object<int, Args..., int>;
};
```

File: 12_variadic_templates/12_02_pack_expansion/12_02_02_braced_initialization.cpp [hunk 11]
```diff
@@ -0,0 +1,21 @@
+// Show nerrow conversion
```
Comment (ivankochin) [inline 12_variadic_templates/12_02_pack_expansion/12_02_02_braced_initialization.cpp]:
```suggestion
// Show narrow conversion
```

File: 12_variadic_templates/12_02_pack_expansion/12_02_04_forwarding.cpp [hunk 12]
```diff
@@ -0,0 +1,14 @@
+template <typename... Args>
+void foo( Args&&... args ) {
+    // ...
+}
+
+template <typename... Args> // Args ~= Arg1, Arg2
+void bar( Args&&... args ) { // args ~= arg1, arg2
```
Comment (ivankochin) [inline 12_variadic_templates/12_02_pack_expansion/12_02_04_forwarding.cpp]:
```suggestion
template <typename... Args> // Arg1, Arg2
void bar( Args&&... args ) { // arg1, arg2
```

File: 12_variadic_templates/12_02_pack_expansion/12_02_06_non_type_template.cpp [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+struct Object {
+
+    template <Args... NonTypeTemplate> // template <int NTT1, std::size_t NTT2>
+    struct Widget {};
+
+};
+
+int main() {
+    Object<int, std::size_t> obj;
+}
```
Comment (ivankochin) [inline 12_variadic_templates/12_02_pack_expansion/12_02_06_non_type_template.cpp]:
```suggestion
template <typename... Args>
struct Object {
    template <Args... NonTypeTemplate>
    struct Widget {};
};

int main() {
    Object<int, std::size_t> obj; // Object::Widget<int NTT1, std::size_t NTT2>
}
```

File: 12_variadic_templates/12_02_pack_expansion/12_02_07_difficult_cases.cpp [hunk 14]
```diff
@@ -0,0 +1,51 @@
+// simultaneous expansion
```
Comment (ivankochin) [inline 12_variadic_templates/12_02_pack_expansion/12_02_07_difficult_cases.cpp]:
```suggestion
// Simultaneous expansion
```

File: 12_variadic_templates/12_02_pack_expansion/12_02_07_difficult_cases.cpp [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+template <typename... Args1>
+struct object {
+
+    template <typename... Args2>
+    struct nested_object : base_class<std::map<Args1, Args2>...> {
+
+        using base_type = base_class<std::map<Args1, Args2>...>;
+    };
+
+};
```
Comment (ivankochin) [inline 12_variadic_templates/12_02_pack_expansion/12_02_07_difficult_cases.cpp]:
```suggestion
template <typename... Args1>
struct object {
    template <typename... Args2>
    struct nested_object : base_class<std::map<Args1, Args2>...> {
        using base_type = base_class<std::map<Args1, Args2>...>;
    };
};
```

File: 12_variadic_templates/12_02_pack_expansion/12_02_07_difficult_cases.cpp [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+template <typename... Args>
+void print();
+
+void type_printer() {
+    using object_type = object<int, float, double>;
+
+    using nested_type = typename object_type::nested_object<std::string, int, short>;
+
+    using nested_base_type = typename nested_type::base_type;
+}
```
Comment (ivankochin) [inline 12_variadic_templates/12_02_pack_expansion/12_02_07_difficult_cases.cpp]:
```suggestion
void type_printer() {
    using object_type = object<int, float, double>;
    using nested_type = typename object_type::nested_object<std::string, int, short>;
    using nested_base_type = typename nested_type::base_type;
}
```

File: 12_variadic_templates/12_02_pack_expansion/12_02_07_difficult_cases.cpp [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+
+// nested expansion
+template <typename... Args>
+int foo(Args... args) {
```
Comment (ivankochin) [inline 12_variadic_templates/12_02_pack_expansion/12_02_07_difficult_cases.cpp]:
```suggestion
int foo( Args... args ) {
```

File: 12_variadic_templates/12_04_recursive_cut_off/12_04_00_recursive_cut_off.cpp [hunk 18]
```diff
@@ -0,0 +1,25 @@
+// Describe the revursive cut off for sum
```
Comment (ivankochin) [inline 12_variadic_templates/12_04_recursive_cut_off/12_04_00_recursive_cut_off.cpp]:
```suggestion
// Describe the recursive cut off for sum
```

File: 12_variadic_templates/12_04_recursive_cut_off/12_04_00_recursive_cut_off.cpp [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+// Show the solution for execute
```
Comment (ivankochin) [inline 12_variadic_templates/12_04_recursive_cut_off/12_04_00_recursive_cut_off.cpp]:
```suggestion
// Show the solution for `execute`
```

**Pull request merged**